numpy/_core/src/multiarray/scalartypes.c.src

```diff
@@ -3904,45 +3904,16 @@ static inline npy_hash_t
  * #lname = datetime, timedelta#
  * #name = Datetime, Timedelta#
  */
-#if NPY_SIZEOF_HASH_T==NPY_SIZEOF_DATETIME
 static npy_hash_t
 @lname@_arrtype_hash(PyObject *obj)
 {
-    npy_hash_t x = (npy_hash_t)(PyArrayScalar_VAL(obj, @name@));
-    if (x == -1) {
-        x = -2;
-    }
-    return x;
+    PyArray_Descr *dtype = PyArray_DescrFromScalar(obj);
+    PyArray_DatetimeMetaData *meta = get_datetime_metadata_from_dtype(dtype);
+    return @lname@_hash(meta, PyArrayScalar_VAL(obj, @name@));
```

Comment:
The check for `NaT` in this function is fair, but the result value is incorrect.  We must return the object identity for `NaT` values.

But I suppose one can argue that is yet a different bug...

Reply:
Why must we return the object identity for NaT values? I must be missing something.

Reply:
It's a hash collision safety thing:  Same as Python `float("nan")` values.

Reply:
Ah right, since NaNs don't compare "normally" ( `float('nan') != float('nan')` ) we can always return a different hash for each NaN.
Interestingly `hash(float('inf')) == 314159`, since comparisons work "normally" for infinity.
I'm pretty sure that historically numpy NaTs compared the "normal" way, unlike NaNs.
But now they compare the "NaN way" (maybe commit c8fc0e80e4fa2ae55a6461d01bf195471e526188 ? )
I'll go make it return the object identity...